File: app/src/main/java/com/robifr/ledger/ui/main/customer/recycler/CustomerListMenu.java
/**
 * Copyright (c) 2024 Robi
 *
 * Ledger is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * Ledger is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
 * GNU General Public License for more details.
 *
 * You should have received a copy of the GNU General Public License
 * along with Ledger. If not, see <https://www.gnu.org/licenses/>.
 */

package com.robifr.ledger.ui.main.customer.recycler;

import android.view.View;
import androidx.annotation.NonNull;
import com.google.android.material.bottomsheet.BottomSheetDialog;
import com.robifr.ledger.R;
import com.robifr.ledger.databinding.CustomerCardDialogMenuBinding;
import com.robifr.ledger.ui.BackStack;
import com.robifr.ledger.ui.editcustomer.EditCustomerFragment;
import com.robifr.ledger.ui.main.customer.CustomerFragment;
import java.util.Objects;

public class CustomerListMenu implements View.OnClickListener {
  @NonNull private final CustomerFragment _fragment;
  @NonNull private final CustomerListHolder _holder;
  @NonNull private final CustomerCardDialogMenuBinding _dialogBinding;
  @NonNull private final BottomSheetDialog _dialog;

  public CustomerListMenu(@NonNull CustomerFragment fragment, @NonNull CustomerListHolder holder) {
    this._fragment = Objects.requireNonNull(fragment);
    this._holder = Objects.requireNonNull(holder);
    this._dialogBinding = CustomerCardDialogMenuBinding.inflate(this._fragment.getLayoutInflater());
    this._dialog =
        new BottomSheetDialog(this._fragment.requireContext(), R.style.BottomSheetDialog);

    this._dialog.setContentView(this._dialogBinding.getRoot());
  }

  @Override
  public void onClick(@NonNull View view) {
    Objects.requireNonNull(view);

    switch (view.getId()) {
      case R.id.editButton -> {
        if (this._holder.boundCustomer().id() == null) return;

        final EditCustomerFragment editCustomerFragment =
            (EditCustomerFragment)
                new EditCustomerFragment.Factory(this._holder.boundCustomer().id())
                    .instantiate(
                        this._fragment.requireContext().getClassLoader(),
                        EditCustomerFragment.class.getName());

        if (this._fragment.requireActivity() instanceof BackStack navigation
            && navigation.currentTabStackTag() != null) {
          navigation.pushFragmentStack(
              navigation.currentTabStackTag(),
              editCustomerFragment,
              EditCustomerFragment.class.toString());
          this._dialog.dismiss();
        }
      }

      case R.id.deleteButton -> {
        this._fragment.customerViewModel().deleteCustomer(this._holder.boundCustomer());
        this._dialog.dismiss();
      }
    }
  }

  public void openDialog() {
    this._dialogBinding.editButton.setOnClickListener(this);
    this._dialogBinding.deleteButton.setOnClickListener(this);
    this._dialog.show();
  }
}

File: app/src/main/java/com/robifr/ledger/ui/customer/recycler/CustomerListMenu.kt
/**
 * Copyright 2024 Robi
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

package com.robifr.ledger.ui.customer.recycler

import android.os.Bundle
import android.view.LayoutInflater
import androidx.core.text.HtmlCompat
import androidx.navigation.findNavController
import com.google.android.material.bottomsheet.BottomSheetBehavior
import com.google.android.material.bottomsheet.BottomSheetDialog
import com.google.android.material.dialog.MaterialAlertDialogBuilder
import com.robifr.ledger.R
import com.robifr.ledger.data.model.CustomerModel
import com.robifr.ledger.databinding.CustomerCardDialogMenuBinding
import com.robifr.ledger.ui.editcustomer.EditCustomerFragment

class CustomerListMenu(
    holder: CustomerListHolder,
    customers: () -> List<CustomerModel>,
    onDeleteCustomer: (CustomerModel) -> Unit,
    customerIndex: () -> Int
) {
  private val _dialogBinding: CustomerCardDialogMenuBinding =
      CustomerCardDialogMenuBinding.inflate(LayoutInflater.from(holder.itemView.context)).apply {
        editButton.setOnClickListener {
          val customerId: Long = customers()[customerIndex()].id ?: return@setOnClickListener
          holder.itemView
              .findNavController()
              .navigate(
                  R.id.editCustomerFragment,
                  Bundle().apply {
                    putLong(
                        EditCustomerFragment.Arguments.INITIAL_CUSTOMER_ID_TO_EDIT_LONG.key(),
                        customerId)
                  })
          _dialog.dismiss()
        }
        deleteButton.setOnClickListener {
          MaterialAlertDialogBuilder(holder.itemView.context)
              .setTitle(
                  HtmlCompat.fromHtml(
                      holder.itemView.context.getString(R.string.customer_cardMenu_deleteWarning),
                      HtmlCompat.FROM_HTML_MODE_LEGACY))
              .setMessage(
                  HtmlCompat.fromHtml(
                      holder.itemView.context.getString(
                          R.string.customer_cardMenu_deleteWarning_description,
                          customers()[customerIndex()].name),
                      HtmlCompat.FROM_HTML_MODE_LEGACY))
              .setNegativeButton(R.string.action_delete) { _, _ ->
                onDeleteCustomer(customers()[customerIndex()])
                _dialog.dismiss()
              }
              .setPositiveButton(R.string.action_cancel) { _, _ -> }
              .show()
        }
      }
  private val _dialog: BottomSheetDialog =
      BottomSheetDialog(holder.itemView.context, R.style.BottomSheetDialog).apply {
        behavior.state = BottomSheetBehavior.STATE_EXPANDED
        setContentView(_dialogBinding.root)
      }

  fun openDialog() {
    _dialog.show()
  }
}
